### PR TITLE
Dufour channels through subgraph

### PR DESCRIPTION
**Pull request by jeandemeusy**
With this PR, Dufour accounts and channels can be queried with a subgraph call. For a given account, incoming and outgoing channels are listed along side with their respective count. 
The total balance in outgoing channels is also linked to the account.

### CODE REVIEW

File: subgraph/packages/subgraph-dufour-channel/schema.graphql [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  balance: BigDecimal!
+}
+
+enum ChannelStatus {
+  CLOSED
+  OPEN
+  PENDING_TO_CLOSE
+}
+
+type Ticket @entity {
```
Comment (QYuQianchen) [inline subgraph/packages/subgraph-dufour-channel/schema.graphql:27]: One field is missing: "indexOffset". Each ticket redemption can redeem a group of tickets. The size of the group is the `indexOffset`. This value can be found as parameters in the ticket redemption txn

File: subgraph/packages/subgraph-dufour-channel/src/channels.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    channel.redeemedTicketCount = channel.redeemedTicketCount.plus(oneBigInt())
+    channel.ticketEpoch = ticket.ticketEpoch
+    channel.ticketIndex = ticket.ticketIndex
+
+    channel.save()
+}
+
+export function handleRedeemTicketCall(call: RedeemTicketCall): void {
```
Comment (QYuQianchen) [inline subgraph/packages/subgraph-dufour-channel/src/channels.ts:126]: will subgraph also index failed calls? or should the failed calls handled by the subgraph creator?

**Review by QYuQianchen — CHANGES_REQUESTED** [review]

File: subgraph/packages/subgraph-dufour-channel/schema.graphql [hunk 3]
```diff
@@ -0,0 +1,57 @@
+type Account @entity {
```
[the remark this answers is not in the document]
Comment (jeandemeusy) [reply subgraph/packages/subgraph-dufour-channel/schema.graphql:1]: Done.

File: subgraph/packages/subgraph-dufour-channel/src/announcements.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        log.error("Revoke announcement for non-existing account", [])
+        return
+    }
+
+    account.hasAnnounced = false
+    account.save()
+}
+
+export function handleKeyBinding(event: KeyBinding): void {
+    let accountId = "foo"
```
[the remark this answers is not in the document]
Comment (jeandemeusy) [reply subgraph/packages/subgraph-dufour-channel/src/announcements.ts]: Error.

File: subgraph/packages/subgraph-dufour-channel/src/channels.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    let channel = Channel.load(channelId)
+
+
+    if (channel == null) {
+        log.error("Decrease balance for non-existing channel", [])
+        return
+    }
+
+    channel.balance = convertEthToDecimal(event.params.newBalance)
+    channel.lastOpenedAt = event.block.timestamp
```
[the remark this answers is not in the document]
Comment (jeandemeusy) [reply subgraph/packages/subgraph-dufour-channel/src/channels.ts]: Done.

File: subgraph/packages/subgraph-dufour-channel/src/channels.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    let channel = Channel.load(channelId)
+
+
+    if (channel == null) {
+        log.error("Increase balance for non-existing channel", [])
+        return
+    }
+
+    channel.balance = convertEthToDecimal(event.params.newBalance)
+    channel.lastOpenedAt = event.block.timestamp
```
[the remark this answers is not in the document]
Comment (jeandemeusy) [reply subgraph/packages/subgraph-dufour-channel/src/channels.ts]: Done.

File: subgraph/packages/subgraph-dufour-channel/schema.graphql (hunk 1 above)
Comment (jeandemeusy) [reply subgraph/packages/subgraph-dufour-channel/schema.graphql:27]: indexOffset is not in the ticket redemption txn, but in the ticket redeem call.. Or I cannot find it.

File: subgraph/packages/subgraph-dufour-channel/src/channels.ts (hunk 2 above)
Comment (jeandemeusy) [reply subgraph/packages/subgraph-dufour-channel/src/channels.ts:126]: As it's the only place where I manage to find many of the parameters, I guess we need to listen to the redeemticket call. However, this call doesn't track if the call fails, or I cannot find it.

File: subgraph/packages/subgraph-dufour-channel/schema.graphql [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  id: ID!
+  "account's public key"
+  publicKey: Bytes
+  multiaddr: [String!]!
+  "channels where the Account is the source"
+  fromChannels: [Channel!]! @derivedFrom(field: "source")
+  "channels where the Account is the destination"
+  toChannels: [Channel!]! @derivedFrom(field: "destination")
+  "number of outgoing channels"
+  fromChannelsCount: BigInt!
```
Comment (QYuQianchen) [inline subgraph/packages/subgraph-dufour-channel/schema.graphql:12]: `fromChannelsCount`, `toChannelsCount` and `balance` in `Account` were not updated when in respective Channels events. When querying an "account" entity with a node address, it should return the total number of outgoing channels in `fromChannelsCount`, the total number of incoming channels in `toChannelsCount` and the sum of stakes in outgoing channels in `balance`

**Review by QYuQianchen — CHANGES_REQUESTED** [review]

File: subgraph/packages/subgraph-dufour-channel/schema.graphql (hunk 7 above)
Comment (jeandemeusy) [reply subgraph/packages/subgraph-dufour-channel/schema.graphql:12]: Thanks ! The `fromChannelCount` and `toChannelCount` were already updated on channel opening and closing. The balance wasn't indeed.
Question: when a channel is opened, the initial amount is not in the `ChannelOpened` event. Should I retrieve it from hash also, or is a ChannelBalanceIncreased event also emitted on channel opening ?

**Pull request merged**